cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -398,29 +401,46 @@ var (
 )
 
 func isCoreDNSVersionSupported(client clientset.Interface) (bool, error) {
-	isValidVersion := true
-	coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
-		context.TODO(),
-		metav1.ListOptions{
-			LabelSelector: "k8s-app=kube-dns",
-		},
-	)
-	if err != nil {
-		return false, errors.Wrap(err, "unable to list CoreDNS pods")
-	}
+	passed := true
+	var lastError error
+	err := wait.PollImmediate(constants.APICallRetryInterval, 10 * time.Second, func() (bool, error) {
+		coreDNSPodList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
+			context.TODO(),
+			metav1.ListOptions{
+				LabelSelector: "k8s-app=kube-dns",
+				FieldSelector: "status.phase=Running",
+			},
+		)
+		if err != nil {
+			lastError = err
+			return false, nil
+		}
 
-	for _, pod := range coreDNSPodList.Items {
-		imageID := imageDigestMatcher.FindStringSubmatch(pod.Status.ContainerStatuses[0].ImageID)
-		if len(imageID) != 2 {
-			return false, errors.Errorf("unable to match SHA256 digest ID in %q", pod.Status.ContainerStatuses[0].ImageID)
+		if len(coreDNSPodList.Items) <= 0 {
```

Comment:
This would only ensure that at least one pod is running, there could be other coredns pods in other states.
So, we would only be checking versions on the running pods.

I dont know how important this distinction is.  Usually all of the pods would be running the same version... the exception being if there is a rollout in progress.

Reply:
i think we should just remove this check.

Reply:
> I dont know how important this distinction is. Usually all of the pods would be running the same version... the exception being if there is a deployment rollout in progress (with image change).

IMO, I agree with this point of view, but the original implementation is to check that there is, whether it needs to be modified?

Reply:
i have never seem a LIST request return < 0 items. is that possible?

Reply:
> i think we should just remove this check and assume the number of pods we go from the above LIST call is what we want to ImageID-check against.

The list call has a field selector for `status.phase=Running`. So, if a Pod is in a state other that "Running", it wont be in the list returned, i.e. we skip the check for all pending Pods.

Reply:
yes, i think the latest version of the PR where the list with the selector being inside the poll, should hopefully get all Pods as running and successfully get the digests from them.

Reply:
Oh OK, so the idea is to _only_ check versions of Running Pods, and not to check versions of any CoreDNS Pods in any other states. I was thinking we wanted to wait for _all_ pods to be non-pending before checking.

Reply:
actually, that makes more sense. IIRC @rajansandeep mentioned that with auto-scaling there could be ~100 of coredns pods in larger clusters, that should not be a problem, but not sure if the 10 second check will be sufficient for all cases.

@zouyee we could loop over all coredns pods and wait until they are Running (max=10 seconds). only then proceed with the checks.